File: leetcode/3Sum.java
import java.util.*;

public class Solution {
    public static ArrayList<Integer> cloneList(ArrayList<Integer> list) {
        ArrayList<Integer> clone = new ArrayList<Integer>(list.size());
        for (Integer item: list) clone.add(item);
        return clone;
    }
    public ArrayList<ArrayList<Integer>> threeSum(int[] num) {
        Arrays.sort(num);
        HashMap<Integer, Integer> single = new HashMap<Integer, Integer>();
        for (int i = 0; i < num.length; ++i) {
            if (single.containsKey(num[i])) {
                int old = single.get(num[i]);
                single.put(num[i], old + 1);
            } else {
                single.put(num[i], 1);
            }
        }
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        HashSet<ArrayList<Integer>> pair = new HashSet<ArrayList<Integer>>();
        for (int i = 0; i < num.length; ++i)
            for (int j = i + 1; j < num.length; ++j) {
                ArrayList<Integer> x = new ArrayList<Integer>();
                x.add(num[i]);
                x.add(num[j]);
                if (!pair.contains(x)) {
                    int y = 0 - num[i] - num[j];
                    if (y < num[j]) continue;
                    int ct = 1;
                    if (y == num[i]) ct++;
                    if (y == num[j]) ct++;
                    pair.add(cloneList(x));
                    if (single.containsKey(y)) {
                        int z = single.get(y);
                        if (z >= ct) {
                            x.add(y);
                            ret.add(cloneList(x));
                        }
                    }
                    
                }
            }
        return ret;
    }
}

File: leetcode/3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        sort(num.begin(), num.end());
        map<int, int> counter;
        for (int i = 0; i < num.size(); ++i)
            if (i == 0 || num[i] != num[i - 1])
                counter[num[i]] = 1;
            else
                counter[num[i]]++;
        vector<vector<int> > triples;
        for (int i = 0; i < num.size(); ++i)
            if (i == 0 || num[i] != num[i - 1])
                for (int j = i + 1; j < num.size(); ++j)
                    if (j == i + 1 || num[j] != num[j - 1]) {
                        int third = -(num[i] + num[j]);
                        if (third >= num[j] && // Forgot this at first
                            counter.find(third) != counter.end()) {
                            int c = counter[third];
                            if (third == num[i])
                                c--;
                            if (third == num[j])
                                c--;
                            if (c > 0) {
                                vector<int> t;
                                t.push_back(num[i]);
                                t.push_back(num[j]);
                                t.push_back(third);
                                triples.push_back(t);
                            }
                        }
                    }
        return triples;
    }
};
